debugger/symbols.ml

```diff
@@ -37,6 +37,18 @@ let events_by_module =
 let all_events_by_module =
   (Hashtbl.create 17 : (string, debug_event list) Hashtbl.t)
 
+let partition_modules evl =
+  let rec partition_modules' ev evl =
+    match evl with
+      [] -> [ev],[]
+    | ev'::evl ->
+       let evl,evll = partition_modules' ev' evl in
+       if ev.ev_module = ev'.ev_module then ev::evl,evll else [ev],evl::evll
+  in
+  match evl with
+    [] -> []
+  | ev::evl -> let evl,evll = partition_modules' ev evl in evl::evll
```

Comment:
If I understand correctly, this code assume that event of the same module are adjacent. I don't understand the code to know if that is always correct. Otherwise, the following versions do not assume this:

```
let SMap = Map.Make(String) in
let add ev map =
  let evl = match SMap.find ev.ev_module with
    | exception Not_found -> [ev]
    | evl -> ev::evl in
  SMap.add ev.ev_module evl map in
let map = List.fold_right add evl in
List.map snd (SMap.bindings map)
```

```
let keys = Hashtbl.create 10 in
let map = Hashtbl.create 10 in
let add ev =
  Hashtbl.replace keys ev.ev_module ();
  Hashtbl.add map ev.ev_module ev
in
List.iter add evl;
Hashtbl.fold (fun k () evll -> Hashtbl.find_all map k :: evll) keys []
```